HW1/P5/cc.py

```diff
@@ -0,0 +1,132 @@
+#import findspark
+#findspark.init()
+#import pyspark
+import time
+start = time.time()
+from pyspark import SparkContext, SparkConf
+conf = SparkConf().setAppName("app").setMaster("local")
+sc = SparkContext(conf=conf)
+
+links = sc.textFile('s3://Harvard-CS205/wikipedia/links-simple-sorted.txt')
+#links = sc.textFile('links.txt')
+start = time.time()
+
+def expand_values((k, v)):
+    res = list()
+    for i in v:
+        res.append((k, i))
+    return res
+
+def expand_values_union((k, v)):
+    res = list()
+    for i in v:
+        res.append((i, k))
+    return res
+
+def reverse_edges(node):
+    edges = []
+    for neighbor in node[1]:
+        edges.append((neighbor, set([node[0]])))
+    return edges
+
+
+def pre_process(links_rdd):
+    rdd = links_rdd
+    rdd1 = rdd.flatMap(expand_values)
+    rdd2 = rdd.flatMap(expand_values_union)
+    rdd_intersect = rdd1.intersection(rdd2).groupByKey()
+    rdd_union = rdd1.union(rdd2).groupByKey()
+    return rdd_intersect, rdd_union
+
+def links_mapper(w):
+    if ': ' not in w:
+        return -1, -1
+    w = w.split(': ')
+    index = int(w[0])
+    neighbors = []
+    for item in w[1].split(' '):
+        if len(item) > 0:
+            neighbors.append(int(item))
+    return index, neighbors
+
+def dataStruCons((k, v)):
+    return (k, (False, v, max_distance, max_cid))
+
+max_distance = 10000000
+max_cid = -1
+
+links_RDD = links.map(links_mapper).partitionBy(4).cache()
+links_intersect, links_union = pre_process(links_RDD)
+
+sym_graph_RDD = links_union.map(dataStruCons).cache()
+revRDD = links_RDD.flatMap(reverse_edges).reduceByKey(lambda x,y: x.union(y)).partitionBy(4)
+asym_graph_RDD = links_RDD.join(revRDD).map(lambda (k,v): (k, (False, (set(v[0]).intersection(v[1])),max_distance,max_cid)))
+
+def bfs_one_node(origin_index, _rdd):
```

Comment:
Yeah I don't think this will ever work on the amazon dataset...there are too many connected components. In the directed case, there are 10s of thousands of unconnected pieces...so the script would take more than 10,000 iterations to complete, which is basically impossible. I will copy what I did.

I have implemented an algorithm that will compute the number of connected components
in O(D) iterations; I don't think it converges in O(log(D)), but maybe I somehow got O(log(D)) unknowingly. The algorithm works in a couple of steps:

1. Assign each node in the graph a unique integer ID
2. Pass the ID of each parent to their children
3. If the ID passed to the child is smaller than the child's current ID, the child adopts the new ID
4. Goto 2, repeat until the number of unique ID's converges to a constant value.

The maximum amount of time it will take for one node to see the effect of another node is O(D) as the maximum number of nodes between two nodes is D and information propagates along one node per iteration.